File: cpp/cowjump.cpp
#include <bits/stdc++.h>


using namespace std;

using ll = long long;

struct Point {
    ll x, y;
};

struct Segment {
    Point a,b;
};

enum EventType {
    Begin,
    End
};

struct Event {
    Point* p;
    EventType e;
    int id;

    Event(Point* p_, EventType e_, int id_)
            : p(p_), e(e_), id(id_) {}
};
long double y_at(Segment& s, ll x) {
    if (s.a.x == s.b.x) return static_cast<long double>(s.a.y);
    return static_cast<long double>(s.a.y) + static_cast<long double>(s.b.y-s.a.y) * static_cast<long double>(x - s.a.x) / static_cast<long double>(s.b.x - s.a.x);
}

struct SegmentComparator {
    std::vector<Segment>* segments;
    long long* sweep_x;

    bool operator()(int id1, int id2) const {
        if (id1 == id2) return false;
        long double y1 = y_at((*segments)[id1], *sweep_x);
        long double y2 = y_at((*segments)[id2], *sweep_x);
        if (y1 != y2) return y1 < y2;
        return id1 < id2;
    }
};

ll orient(Point& a, Point& b, Point& c) {
    return (b.x - a.x) * (c.y - a.y) -
           (b.y - a.y) * (c.x - a.x);

}

bool on_segment(Point& a, Point& b, Point& p) {
    return min(a.x, b.x) <= p.x && p.x <= max(a.x, b.x) &&
           min(a.y, b.y) <= p.y && p.y <= max(a.y, b.y);
}

bool intersect (Segment& s1, Segment& s2) {
    auto& [a,b] = s1;
    auto& [c, d] = s2;
    auto o1 = orient(a, b, c);
    auto o2 = orient(a, b, d);
    auto o3 = orient(c, d, a);
    auto o4 = orient(c, d, b);

    if (((o1 > 0 && o2 < 0) || (o1 < 0 && o2 > 0)) && 
        ((o3 > 0 && o4 < 0) || (o3 < 0 && o4 > 0))) return true;

    if (o1 == 0 && on_segment(a, b, c)) return true;
    if (o2 == 0 && on_segment(a, b, d)) return true;
    if (o3 == 0 && on_segment(c, d, a)) return true;
    if (o4 == 0 && on_segment(c, d, b)) return true;
    return false;
}

int main() {
  freopen("cowjump.in", "r", stdin);
  freopen("cowjump.out", "w", stdout);
  ios::sync_with_stdio(false);
  cin.tie(nullptr);

  int n;

  while (cin >> n) {
    vector<Event> es;
    vector<Segment> segments;
    segments.resize(n);
    es.reserve(2*n);
    for (int i = 0; i < n; i++) {
        cin >> segments[i].a.x >> segments[i].a.y >> segments[i].b.x >> segments[i].b.y;
        if (segments[i].a.x > segments[i].b.x)
            swap(segments[i].a, segments[i].b);

        es.emplace_back(&segments[i].a, EventType::Begin, i);
        es.emplace_back(&segments[i].b, EventType::End, i);
    }
    sort(es.begin(), es.end(), [](const auto& a, const auto& b) {
        const auto& [pa, ea, ida] = a;
        const auto& [pb, eb, idb] = b;
        if (pa->x != pb->x)
            return pa->x < pb->x;
        if (ea != eb)
            return ea < eb;
        if (pa->y != pb->y)
            return pa->y < pb->y;
        return ida < idb;
    });
    ll curr_sweep_x;
    SegmentComparator cmp {&segments, &curr_sweep_x};
    set<int, SegmentComparator> active (cmp);

    int ans_id = -1;
    auto check_and_set = [&](int id1, int id2) {
        int count1 = 0, count2 = 0;
        for (int i = 0; i < n; i++) {
            if (i != id1 && intersect(segments[id1], segments[i])) count1++;
            if (i != id2 && intersect(segments[id2], segments[i])) count2++;
        }
        if (count1 > 1) ans_id = id1 + 1;
        else if (count2 > 1) ans_id = id2 + 1;
        else ans_id = min(id1, id2) + 1;
    };

    for(auto& [p, e, id] : es){
        curr_sweep_x = p->x;
        if (e == EventType::Begin){
            auto itr = active.insert(id).first;
            auto& s = segments[*itr];
            auto above = next(itr);
            auto below = (itr == active.begin()) ? active.end() : prev(itr);
            if (above != active.end() && intersect(s, segments[*above])){
                check_and_set(id, *above);
                break;
            }
            if (below != active.end() && intersect(s, segments[*below])){
                check_and_set(id, *below);
                break;
            }
        }
        if (e == EventType::End) {
            auto itr = active.find(id);
            if (itr != active.end()) {
                auto above = next(itr);
                auto below = (itr == active.begin()) ? active.end() : prev(itr);
                if (above != active.end() && below != active.end() && intersect(segments[*above], segments[*below])){
                    check_and_set(*above, *below);
                    break;
                }
                active.erase(itr);
            }
        }
    }
    if (ans_id != -1) cout << ans_id << "\n";
  }

}

File: src/bin/cowjump.rs
use std::cell::Cell;
use std::cmp::Ordering;
use std::collections::BTreeSet;
use std::fs::File;
use std::io::{self, Read, Write};
use std::ops::Bound::{Excluded, Unbounded};

#[derive(Copy, Clone, Debug)]
struct Point {
    x: i64,
    y: i64,
}

#[derive(Copy, Clone, Debug)]
struct Segment {
    a: Point,
    b: Point,
}

#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
enum EventType {
    Begin,
    End,
}

#[derive(Copy, Clone, Debug)]
struct Event {
    x: i64,
    y: i64,
    kind: EventType,
    id: usize,
}

#[derive(Copy, Clone)]
struct ActiveSegment<'a> {
    id: usize,
    segments: &'a [Segment],
    sweep_x: &'a Cell<i64>,
}

impl<'a> PartialEq for ActiveSegment<'a> {
    fn eq(&self, other: &Self) -> bool {
        self.id == other.id
    }
}

impl<'a> Eq for ActiveSegment<'a> {}

impl<'a> PartialOrd for ActiveSegment<'a> {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

impl<'a> Ord for ActiveSegment<'a> {
    fn cmp(&self, other: &Self) -> Ordering {
        if self.id == other.id {
            return Ordering::Equal;
        }

        cmp_segments_at(self.segments, self.id, other.id, self.sweep_x.get())
    }
}

fn y_key(seg: &Segment, x: i64) -> (i128, i128) {
    let dx = (seg.b.x - seg.a.x) as i128;
    if dx == 0 {
        return (seg.a.y as i128, 1);
    }

    let dy = (seg.b.y - seg.a.y) as i128;
    let offset = (x - seg.a.x) as i128;
    let num = seg.a.y as i128 * dx + dy * offset;
    (num, dx)
}

fn cmp_segments_at(segments: &[Segment], id1: usize, id2: usize, x: i64) -> Ordering {
    let (num1, den1) = y_key(&segments[id1], x);
    let (num2, den2) = y_key(&segments[id2], x);

    (num1 * den2)
        .cmp(&(num2 * den1))
        .then_with(|| id1.cmp(&id2))
}

fn orient(a: &Point, b: &Point, c: &Point) -> i128 {
    (b.x - a.x) as i128 * (c.y - a.y) as i128 - (b.y - a.y) as i128 * (c.x - a.x) as i128
}

fn on_segment(a: &Point, b: &Point, p: &Point) -> bool {
    a.x.min(b.x) <= p.x && p.x <= a.x.max(b.x) && a.y.min(b.y) <= p.y && p.y <= a.y.max(b.y)
}

fn intersect(s1: &Segment, s2: &Segment) -> bool {
    let a = &s1.a;
    let b = &s1.b;
    let c = &s2.a;
    let d = &s2.b;

    let o1 = orient(a, b, c);
    let o2 = orient(a, b, d);
    let o3 = orient(c, d, a);
    let o4 = orient(c, d, b);

    if ((o1 > 0 && o2 < 0) || (o1 < 0 && o2 > 0)) && ((o3 > 0 && o4 < 0) || (o3 < 0 && o4 > 0)) {
        return true;
    }

    if o1 == 0 && on_segment(a, b, c) {
        return true;
    }
    if o2 == 0 && on_segment(a, b, d) {
        return true;
    }
    if o3 == 0 && on_segment(c, d, a) {
        return true;
    }
    if o4 == 0 && on_segment(c, d, b) {
        return true;
    }
    false
}

fn choose_answer(segments: &[Segment], id1: usize, id2: usize) -> usize {
    let mut count1 = 0;
    let mut count2 = 0;

    for i in 0..segments.len() {
        if i != id1 && intersect(&segments[id1], &segments[i]) {
            count1 += 1;
        }
        if i != id2 && intersect(&segments[id2], &segments[i]) {
            count2 += 1;
        }
    }

    if count1 > 1 {
        id1 + 1
    } else if count2 > 1 {
        id2 + 1
    } else {
        id1.min(id2) + 1
    }
}

fn solve(input: &str) -> String {
    let mut tokens = input.split_whitespace();
    let mut answers = Vec::new();

    while let Some(n_token) = tokens.next() {
        let n: usize = n_token.parse().unwrap();
        let mut segments = Vec::with_capacity(n);
        let mut events = Vec::with_capacity(2 * n);

        for id in 0..n {
            let mut a = Point {
                x: tokens.next().unwrap().parse().unwrap(),
                y: tokens.next().unwrap().parse().unwrap(),
            };
            let mut b = Point {
                x: tokens.next().unwrap().parse().unwrap(),
                y: tokens.next().unwrap().parse().unwrap(),
            };

            if a.x > b.x {
                std::mem::swap(&mut a, &mut b);
            }

            segments.push(Segment { a, b });
            events.push(Event {
                x: a.x,
                y: a.y,
                kind: EventType::Begin,
                id,
            });
            events.push(Event {
                x: b.x,
                y: b.y,
                kind: EventType::End,
                id,
            });
        }

        events.sort_unstable_by(|lhs, rhs| {
            lhs.x
                .cmp(&rhs.x)
                .then_with(|| lhs.kind.cmp(&rhs.kind))
                .then_with(|| lhs.y.cmp(&rhs.y))
                .then_with(|| lhs.id.cmp(&rhs.id))
        });

        let sweep_x = Cell::new(0_i64);
        let mut active: BTreeSet<ActiveSegment> = BTreeSet::new();
        let mut answer = None;

        for event in events {
            sweep_x.set(event.x);
            let key = ActiveSegment {
                id: event.id,
                segments: &segments,
                sweep_x: &sweep_x,
            };

            match event.kind {
                EventType::Begin => {
                    active.insert(key);

                    let above = active.range((Excluded(key), Unbounded)).next().copied();
                    let below = active.range(..key).next_back().copied();

                    if let Some(above) = above
                        && intersect(&segments[event.id], &segments[above.id])
                    {
                        answer = Some(choose_answer(&segments, event.id, above.id));
                        break;
                    }

                    if let Some(below) = below
                        && intersect(&segments[event.id], &segments[below.id])
                    {
                        answer = Some(choose_answer(&segments, event.id, below.id));
                        break;
                    }
                }
                EventType::End => {
                    let above = active.range((Excluded(key), Unbounded)).next().copied();
                    let below = active.range(..key).next_back().copied();

                    if let (Some(above), Some(below)) = (above, below)
                        && intersect(&segments[above.id], &segments[below.id])
                    {
                        answer = Some(choose_answer(&segments, above.id, below.id));
                        break;
                    }

                    active.remove(&key);
                }
            }
        }

        if let Some(answer) = answer {
            answers.push(answer.to_string());
        }
    }

    answers.join("\n")
}

fn main() -> io::Result<()> {
    let mut input = String::new();
    File::open("cowjump.in")?.read_to_string(&mut input)?;

    let output = solve(&input);
    let mut file_output = File::create("cowjump.out")?;
    writeln!(file_output, "{output}")?;
    Ok(())
}
